third_party/move/move-vm/types/src/loaded_data/runtime_access_specifier.rs

```diff
@@ -0,0 +1,536 @@
+pub enum ResourceSpecifier {
+    Any,
+    DeclaredAtAddress(AccountAddress),
+    DeclaredInModule(ModuleId),
+    Resource(StructIdentifier),
+    ResourceInstantiation(StructIdentifier, Vec<Type>),
+}
+
+/// Represents an address specifier.
+#[derive(PartialEq, Eq, PartialOrd, Ord, Hash, Clone, Debug)]
+pub enum AddressSpecifier {
+    Any,
+    Literal(AccountAddress),
+    /// The `Eval` specifier represents a value dependent on a parameter of the
+    /// current function. Once address specifiers are instantiated in a given
+    /// caller context it is replaced by a literal.
+    Eval(AddressSpecifierFunction, LocalIndex),
+}
+
+/// Represents a well-known function used in an address specifier.
+#[derive(PartialEq, Eq, PartialOrd, Ord, Hash, Clone, Copy, Debug)]
+pub enum AddressSpecifierFunction {
+    Identity,
+    SignerAddress,
+    ObjectAddress,
+}
+
+/// A trait representing an environment for evaluating dynamic values in access specifiers.
+pub trait AccessSpecifierEnv {
+    fn eval_address_specifier_function(
+        &self,
+        fun: AddressSpecifierFunction,
+        local: LocalIndex,
+    ) -> PartialVMResult<AccountAddress>;
+}
+
+/// A struct to represent an access.
+#[derive(PartialEq, Eq, PartialOrd, Ord, Hash, Clone, Debug)]
+pub struct AccessInstance {
+    pub kind: AccessKind,
+    pub resource: StructIdentifier,
+    pub instance: Vec<Type>,
+    pub address: AccountAddress,
+}
+
+impl AccessSpecifier {
+    /// Returns true if this access specifier is known to have no accesses. Note that this
+    /// may be under-approximated in the presence of exclusions. That is, if
+    /// `!s.is_empty()`, it is still possible that all concrete accesses fail.
+    pub fn is_empty(&self) -> bool {
+        if let AccessSpecifier::Constraint(incl, _) = self {
+            incl.is_empty()
+        } else {
+            false
+        }
+    }
+
+    /// Specializes the access specifier for the given environment. This evaluates
+    /// `AddressSpecifier::Eval` terms.
+    pub fn specialize(&mut self, env: &impl AccessSpecifierEnv) -> PartialVMResult<()> {
+        match self {
+            AccessSpecifier::Any => Ok(()),
+            AccessSpecifier::Constraint(incls, excls) => {
+                for clause in incls {
+                    clause.specialize(env)?;
+                }
+                for clause in excls {
+                    clause.specialize(env)?;
+                }
+                Ok(())
+            },
+        }
+    }
+
+    /// Returns true if the concrete access instance is enabled.
+    pub fn enables(&self, access: &AccessInstance) -> bool {
+        use AccessSpecifier::*;
+        match self {
+            Any => true,
+            Constraint(incls, excls) => {
+                incls.iter().any(|c| c.enables(access)) && excls.iter().all(|c| !c.enables(access))
+            },
+        }
+    }
+
+    pub fn join(&self, other: &Self) -> Self {
```

Comment:
Also want to call out the potential space complexity here. Since we are going quadratic at every join how big can this joined result be in prod?

Reply:
The good news is that quadratic is the worst case scenario. In `a*c + a*d + b*c + b*d`, it is expected that many products are empty and dropped. Otherwise, we still need to make the math for the worst case scenario, and ideally represent the bound via gas.